query-planner-js/src/utilities/debug.ts

```diff
@@ -0,0 +1,143 @@
+   *
+   * @param values - an array of the values to log.
+   * @param fct - the function to apply to each of the value to convert it to a string to log.
+   */
+  public groupedValues<T>(values: T[], fct: (v: T) => string) {
+    if (!this.enabled) return this;
+
+    this.group();
+    const indentStr = indentString(this.currentIndentLevel);
+    for (const value of values) {
+      console.log(indentStr + '- ' + fct(value));
+    }
+    return this.groupEnd();
+  }
+
+  /**
+   * Logs the entries of the provided map, indented and one per line.
+   *
+   * @param map - a map of keys and values to log.
+   * @param keyFct - the function to apply to each of the key to convert it to a string to log.
+   * @param valueFct - the function to apply to each of the value to convert it to a string to log.
+   */
+  public groupedEntries<K, V>(
+    map: Map<K, V>,
+    keyFct: (k: K) => string,
+    valueFct: (v : V) => string) {
+    if (!this.enabled) return this;
+
+    this.group();
+    const indentStr = indentString(this.currentIndentLevel);
+    for (const [k, v] of map.entries()) {
+      console.log(indentStr + '- ' + keyFct(k) + ': ' + valueFct(v));
+    }
+    return this.groupEnd();
+  }
+
+  /**
+   * Starts a new debug "group"; every following message logged after this called and before the corresponding
+   * `groupEnd()` call will be indented by one more level than is current.
+   *
+   * @param openingMessage - an optional message to log _before_ starting the group. This can typically be used to
+   * describe what the message logged within the group will describe.
+   */
+  public group(openingMessage?: string | (() => string)) {
+    if (this.enabled) {
+      if (openingMessage) {
+        this.log(openingMessage, chalk.blue('‣ '));
+      }
+      this.currentIndentLevel++;
+    }
+    return this;
+  }
+
+  /**
+   * Ends the current (last started) debug "group", thus decreasing the indent level to what it was before the
+   * `group()` call this ends.
+   *
+   * If no group had been started, this is a no-op (and no `closingMessage` is printed in particular).
+   *
+   * @param closingMessage - an optional message to log _after_ ending the group. This can typically be used to
+   * describe what the messages logged within the group yielded.
+   */
+  public groupEnd(closingMessage?: string | (() => string)) {
+    if (!this.enabled || this.currentIndentLevel == 0) {
+      return this;
+    }
+    this.currentIndentLevel--;
+    if (closingMessage) {
+      this.log(closingMessage, chalk.green('⇒ '));
+    }
+    return this;
+  }
+}
+
+function isTrue(str?: string) : boolean {
+  if (!str) {
+    return false;
+  }
+  switch (str.toLocaleLowerCase()) {
+    case "true":
+    case "yes":
+    case "1":
+      return true;
+    default:
+      return false;
+  }
+}
+
+/**
+ * A global `DebugLogger` for use in the query planner.
+ */
+const debug = new DebugLogger(isTrue(process.env.QP_DEBUG));
+export { debug };
```

Comment:
I think it might make sense to keep the initialization of a global instance out of this file, and make that part of `buildQueryPlan`. That way, we could also enable debugging per module.

Reply:
Agreed and made the change. 

I initially put it here because I figured it may be useful to have access to the same instance (so as to keep the indentation) in multiple files (I've added logs in `selectionSetFromFieldSet` in `FieldSet.ts` for instance at one point, though it's not in this patch), and `debug.ts` felt like a "neutral" place.

But it is the logger for debugging "building plans", so it makes more sense to declare it in `buildQueryPlan.ts` and export it if we need to use it in a few dependent places.